### PR TITLE
Service manager/issue 124 bl implementation

### PR DESCRIPTION
**Pull request by LusineHovs**

### CODE REVIEW

File: src/ServiceManager/Travely.ServiceManager.Abstraction/Interfaces/Repositories/IActivityRepository.cs [hunk 1]
```diff
@@ -1,8 +1,12 @@
...
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Travely.ServiceManager.Abstraction.Models.Db;
 
 namespace Travely.ServiceManager.Abstraction.Interfaces
 {
     public interface IActivityRepository : IRepository<Activity>
     {
+        Task<List<Activity>> GetAllActivitiesByAgencyIdAsync(long agencyId);
+        Task<Activity> GetActivityByUniqueKeysAsync(long agencyId, string activityName, long activityTypeId);
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Abstraction/Interfaces/Repositories/IActivityRepository.cs:10]: The method have any argument called "key".  What you mean by "ByUniqueKeys"?

File: src/ServiceManager/Travely.ServiceManager.Abstraction/Interfaces/Repositories/IActivityTypeRepository.cs [hunk 2]
```diff
@@ -1,8 +1,10 @@
-﻿using Travely.ServiceManager.Abstraction.Models.Db;
+﻿using System.Threading.Tasks;
+using Travely.ServiceManager.Abstraction.Models.Db;
 
 namespace Travely.ServiceManager.Abstraction.Interfaces.Repositories
 {
     public interface IActivityTypeRepository : IRepository<ActivityType>
     {
+        Task<ActivityType> GetActivityTypeByUniqueKeysAsync(long agencyId, string activityTypeName);
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Abstraction/Interfaces/Repositories/IActivityTypeRepository.cs:8]: The method have any argument called "key". What you mean by "ByUniqueKeys"?

File: src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityRepository.cs [hunk 3]
```diff
@@ -1,13 +1,29 @@
...
     {
-        public ActivityRepository(ServiceManagerDbContext dbContext)
-            : base(dbContext)
+        private readonly ServiceManagerDbContext _serviceManagerDbContext;
+
+        public ActivityRepository(ServiceManagerDbContext serviceManagerDbContext)
+            : base(serviceManagerDbContext)
+        {
+            _serviceManagerDbContext = serviceManagerDbContext;
+        }
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityRepository.cs:18]: New line below.

File: src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityRepository.cs [hunk 4]
```diff
@@ -1,13 +1,29 @@
...
+        public async Task<List<Activity>> GetAllActivitiesByAgencyIdAsync(long agencyId)
+        {
+            return await _serviceManagerDbContext.Activities.Include(x => x.ActivityType).Where(x => x.ActivityType.AgencyId == agencyId).ToListAsync();
+        }
+
+        public async Task<Activity> GetActivityByUniqueKeysAsync(long agencyId, string activityName, long activityTypeId)
         {
+            return await _serviceManagerDbContext.Activities.Include(x => x.ActivityType).Where(x => x.ActivityType.AgencyId == agencyId && x.Name == activityName && x.ActivityTypeId == activityTypeId).FirstOrDefaultAsync();
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityRepository.cs:26]: Could you split this to multiple lines?

Comment (tadart) [reply src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityRepository.cs:26]: Should this be First or Single?

File: src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityTypeRepository.cs [hunk 5]
```diff
@@ -1,13 +1,22 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Travely.ServiceManager.Abstraction.Interfaces.Repositories;
 using Travely.ServiceManager.Abstraction.Models.Db;
 
 namespace Travely.ServiceManager.DAL.Repositories
 {
     internal class ActivityTypeRepository : Repository<ActivityType>, IActivityTypeRepository
     {
+        private readonly ServiceManagerDbContext _serviceManagerDbContext;
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityTypeRepository.cs:11]: New line below.

Comment (tadart) [reply src/ServiceManager/Travely.ServiceManager.DAL/Repositories/ActivityTypeRepository.cs:11]: Should this be First or Single?

File: src/ServiceManager/Travely.ServiceManager.DAL/Repositories/Repository.cs [hunk 6]
```diff
@@ -37,14 +37,15 @@ public virtual async Task<IEnumerable<TEntity>> GetAllAsync()
...
-        public virtual async Task<TEntity> GetAsync(int entityId)
+        public virtual async Task<TEntity> GetAsync(long entityId)
         {
             return await _dbSet.FindAsync(entityId);
         }
 
         public virtual void Update(TEntity entity)
         {
             _dbSet.Update(entity);
+            _dbContext.SaveChanges();
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.DAL/Repositories/Repository.cs:48]: Do we really want to have SaveChanges() called in Repository if we have UnitOfWork?

File: src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs [hunk 7]
```diff
@@ -1,36 +1,61 @@
...
     {
         protected readonly IUnitOfWork _unitOfWork;
+        protected readonly IMapper _mapper;
 
-        public ActivityManager(IUnitOfWork unitOfWork)
+        public ActivityManager(IUnitOfWork unitOfWork, IMapper mapper)
         {
             _unitOfWork = unitOfWork;
+            _mapper = mapper;
         }
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs:18]: New line below.

File: src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs [hunk 8]
```diff
@@ -1,36 +1,61 @@
...
-        public Task<Activity> CreateActivityAsync(Activity activity)
+        public async Task<Activity> CreateActivityAsync(Activity activity)
         {
-            throw new NotImplementedException();
+            var activityDTO = _mapper.Map<ServiceManagerDb.Activity>(activity);
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs:21]: We create ServiceManagerDb.Activity twice - activityDTO & activityEntity. Probably, first one is not necessary. Please consider to remove it and work directly with activity.

File: src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs [hunk 9]
```diff
@@ -1,36 +1,61 @@
...
-            throw new NotImplementedException();
+            await _unitOfWork.ActivityRepository.DeleteAsync(activityId);
+            await _unitOfWork.SaveAsync();
         }
 
-        public Task<Activity> EditActivityAsync(Activity activity)
+        public Activity EditActivity(Activity activity)
         {
-            throw new NotImplementedException();
+            var activityDTO = _mapper.Map<ServiceManagerDb.Activity>(activity);
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs:56]:
```suggestion
            var activityEntity = _mapper.Map<ServiceManagerDb.Activity>(activity);
```

File: src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs [hunk 10]
```diff
@@ -1,36 +1,61 @@
...
+            _unitOfWork.ActivityRepository.Update(activityDTO);
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs:57]: Should be saved by UnitOfWork to keep the system consistent.

File: src/ServiceManager/Travely.ServiceManager.Abstraction/Interfaces/Repositories/IActivityTypeRepository.cs (hunk 2 above)
Comment (LusineHovs) [reply src/ServiceManager/Travely.ServiceManager.Abstraction/Interfaces/Repositories/IActivityTypeRepository.cs:8]: agree that the methods names' are inappropriate

File: src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs [hunk 11]
```diff
@@ -1,36 +1,62 @@
...
+            if (activityType == null)
+            {
+                activityType = await _unitOfWork.ActivityTypeRepository.CreateAsync(_mapper.Map<ServiceManagerDb.ActivityType>(activity.Type));
+            }
+
+            ServiceManagerDb.Activity activityEntity;
+            if (activityType.Id != 0)
+            {
+                activityEntity = await _unitOfWork.ActivityRepository.GetActivityByNameAndTypeId(activityType.AgencyId, activity.Name, activityType.Id);
+                if (activityEntity != null) { return null; }
```
Comment (tadart) [inline src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs:32]: Please consider to throw an exception here.

File: src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs [hunk 12]
```diff
@@ -1,36 +1,64 @@
...
         }
-        public Task<Activity> CreateActivityAsync(Activity activity)
+
+        public async Task<Activity> CreateActivityAsync(Activity activity)
         {
-            throw new NotImplementedException();
+            var activityType = await _unitOfWork.ActivityTypeRepository.GetActivityTypeAsync(activity.Type.AgencyId, activity.Type.ActivityName);
+            if (activityType == null)
+            {
+                activityType = await _unitOfWork.ActivityTypeRepository.CreateAsync(_mapper.Map<ServiceManagerDb.ActivityType>(activity.Type));
```
Comment (martikyan) [inline src/ServiceManager/Travely.ServiceManager.Service/Managers/ActivityManager.cs:26]: Please use a local var for the map result to make debugging easier

**Review by martikyan — APPROVED** [review]

**Review by tadart — APPROVED** [review]

**Review by martikyan — APPROVED** [review]

**Pull request merged**